### DISCUSSION TITLE
QBJS Paint

### DISCUSSION THREAD

**Discussion by boxgaming** [boxgaming/qbjs#3] [Show and tell]
Here's a simple paint program that demonstrates the ability to use rich HTML controls and events with your QBJS program.  It supports several different drawing tools, a color picker and supports a single level of undo.

[View in QBJS](https://v6p9d9t4.ssl.hwcdn.net/html/5512918/index.html?qbcode=JEaW0BAKptDC5MrJw3Rvb2zALECxuH6MxOjcqtzI30AKhKS3NLoht7c6OTewWxzc5IVmaW1hZ2U2isbS2sLOy7twB7MAvmOnMruktrCzsqAFAgGxAGnwDD4IPAAKWGBAOUfa9iv2+OlZsZkcmF3aW5n5EI2MLm6LHO4jYwubos+kFc3RhcnRYFqK5ujC5Oiy7oVyYWRpdXP/zCU3ViuCU9uTW91c2VNb3bEBlVVsCks0Y8lNb3VzZUJ1dHRv0AbmNgGPlkGqNDK3Kqr5PCEnN7p0N+f///dfx1B1dEltYWdlYtxRoAH3ohGcmVlSW1hZ2VfwAHjwrPfN9rEtd40ADwWK5xU1vdXNlWNAA6GTFcVNb3VzZVnQH3CCmb/9Khnm//WTToBalt/+tFgC55HZGlzYWJsZWSaSEZmFsc2Vv9zDRWxzZW//cTwlICOzC2OrLVcQCIYOjOTKytDC3Minnm//9AsNEZXN0rN//5+GmNLcyznIV5pOognfQWnviCVNlbGVjdGVkQ294CbG9yb//389D03//vp6Hlv//cIyt//7f/r4qK2ObKks0eOGZ4p4lrhv//WLUm//9SmsyccOIE8hTl//9rtZX//1Oe7BcjnNJxcLiyIUei1//95CVW//RWcJok1WikKTKxujC3M7Yyq6Rv//epVm//9ea9QqhTK6mUqa//+87eX//x9oc6NTGNvKOvPtOlpY6SAIS//+CWdV//JZQm1yxGWKjNLY2MrJGYtCv//g1XV//92WCF5Ckd1I521//8Gd9f//I3iDo9sc89I68+260tDowChIy//+DShV//ZJw2NkivwVDaXJjbGUexX//wE2l//96SDBpKmdtM53F//85nEJBYnN0EktMOUsAfRvfA/Ny71an//+NJLIECyqmn0//9xU///1HLJxArr5l9P//bISK3MluT//3uLqf/+2sSlc+FcmbZmdP//e5mRP/1+dOB5ZDM1URYp//7w1xf//ckvPvoDzNnmM+v//vVMEDC3pH6L2sQJberftk///1SSME9pBrwp//6uT///apJYb2UGxEv//othV//97u4v//hdFgOS//+UgkZvcrq5JpdBVG+AsGGpujK4Sg6AM1///3bCzTYhTNibCwL//7shpzK8OlHr//72sq//m+OX/bUmvtqv+2o5V8lol8xBIbY5r7zbpodfeRKbSo0l/VULkUhPbkJ0blVuZG9ve4Vm9aO8mImcVM4xgt5ELETHIAbo5OrK3loWbLdAsztkURTeLiURvbUdldEltYWcWAMAup4jm6PLYynnOFY3Vyc29yjuCvIY3Jvc3NoYWlyeN4kHRG9tRXZlbnSJr0CYCb4ZkNre6ubK2t7synWCRFlfgnN1YnD0Yb28lo6I4MLcytje+jFfQhEb21DcmVhdGUL94gmRpdgX28u8gIhubgwtwVwmmSu4ao3t7YpIB1Mb3qKEAGvYK5srYysbo91Svt5/oPbF95opPGIht7Y3uTx6be8eK8gJrmI0tzg6ujwlz7e7p93A3R5cGWKqVYjG3tje5ON6USKkVx2gCM7RWZmZmZmZnF9fO0csC46wrE6ujo3txXSz3BHDVW5kb3AL7iavGap8IzNjewukd1wBHJpZ2h0fXyhu4LdCxnBhZGRpbmeS+0QBqLYLg8XeG2iFfQwG4ku+X3KU88LaCMbY0sbWi3AMbQq+1AGcZrCHRleHRBbGlnbpIXfhtjKzOgV9qGaRKTCO7SyOjRLECAA14l9+MDjDpFbWFyZ2luljBaANhdXRvGvv55yi2uWRtJxIvjUWJFnyWZvbnRGYW1pbHmSVEhEFyaWFsNGUhoZWx2ZXRpY2E1lA3NhbnNC+cRzZXJpZir49hmYSYdmb250U2l6ZY5IgkRcyABwy4vgsraVfPsObCRCsTe5MjK5RyjhnXviObe2NLJZACFEDL59h+bPYliYWNrZ3JvdW5kYTHMnKdjyfEy+mbEupwS7Mrk6NLGwtiC2CiE0s7dHMcwra0sjI2MplX3jkpLc0uio3t7YmNKUAuboqRfy+olE2T6vszot0vyW1hcmdpblJpZ2hIAHSSZPCl9HX2t0s6z0k5xEEL7s8XAlVKclB6hOje4IXxKgshW9wdGlvbjuFHQyJC+U0Hqn0OOh3sXzywMXCLHYGLF9PtCrnMjtTKFGr7labwU4O1LNXwNrhCTpLXFuBqoMxX1PUdGdW5jdGlvboiI+vuGeCTUAZywJAYixyAY19lbFPD44q+bIBI5XqkpTC7MLmxuTS4Oi/7Lih8O4MLk5sqS3OiZsPPnFc3Vic3RyTcJTQAMki38HSX/NrFNCw/zxj2EMUx1IsiIJL/mzjkuwZniWabxSVTLMiaa+aVIuy/6+RB0cRnhKSOhNABZnFk4yrFNuvh)
```vb
Dim Shared tool, cp, btnUndo
InitControls

Dim fimage, cimage
fimage = _NewImage(640, 400)
cimage = _NewImage(640, 400)

Dim Shared drawing, lastX, lastY, startX, startY, radius

Sub OnMouseMove
    If _MouseButton(1) Then
        If Not drawing Then
            _PutImage , cimage, fimage
            _FreeImage cimage
            cimage = _NewImage(640, 400)
            lastX = _MouseX
            lastY = _MouseY
            startX = lastX
            startY = lastY
            drawing = -1
            btnUndo.disabled = false
        Else
            If tool.value = "Freehand" Then
                _Dest cimage
                Line (lastX, lastY)-(_MouseX, _MouseY), SelectedColor
                lastX = _MouseX
                lastY = _MouseY
                _Dest 0
                
            ElseIf tool.value = "Line" Then
                _FreeImage cimage
                cimage = _NewImage(640, 400)
                _Dest cimage
                Line (startX, startY)-(_MouseX, _MouseY), SelectedColor
                _Dest 0

            ElseIf tool.value = "Rectangle" Then
                _FreeImage cimage
                cimage = _NewImage(640, 400)
                _Dest cimage
                Line (startX, startY)-(_MouseX, _MouseY), SelectedColor, B
                _Dest 0

            ElseIf tool.value = "Filled Rectangle" Then
                _FreeImage cimage
                cimage = _NewImage(640, 400)
                _Dest cimage
                Line (startX, startY)-(_MouseX, _MouseY), SelectedColor, BF
                _Dest 0

            ElseIf tool.value = "Circle" Then
                _FreeImage cimage
                cimage = _NewImage(640, 400)
                If Abs(_MouseX - startX) > Abs(_MouseY - startY) Then
                    radius = Abs(_MouseX - startX)
                Else
                    radius = Abs(_MouseY - startY)
                End If
                _Dest cimage
                Circle (startX, startY), radius, SelectedColor
                _Dest 0

            ElseIf tool.value = "Filled Circle" Then
                _FreeImage cimage
                cimage = _NewImage(640, 400)
                If Abs(_MouseX - startX) > Abs(_MouseY - startY) Then
                    radius = Abs(_MouseX - startX)
                Else
                    radius = Abs(_MouseY - startY)
                End If
                _Dest cimage
                Dim r
                For r = 0 To radius Step .3
                    Circle (startX, startY), r, SelectedColor
                Next r
                _Dest 0
            End If
        End If
    Else 
        drawing = 0
    End If
    Cls
    _PutImage , fimage
    _PutImage , cimage
End Sub    

Sub OnBtnUndo
    _FreeImage cimage
    cimage = _NewImage(640, 400)
    btnUndo.disabled = true
    OnMouseMove
End Sub

Sub InitControls
    DomGetImage(0).style.cursor = "crosshair"
    DomEvent DomGetImage(0), "mousemove", sub_OnMouseMove
    
    Dim panel
    panel = DomCreate("div")
    DomCreate "span", panel, "Tool: "
    tool = DomCreate("select", panel)
    DomCreate "span", panel, "Color: "
    cp = DomCreate("input", panel)
    cp.type = "color"
    cp.value = "#ffffff"
    
    btnUndo = DomCreate("button", panel, "Undo")
    btnUndo.style.float = "right"
    btnUndo.style.padding = "5px 10px"
    btnUndo.disabled = true
    DomEvent btnUndo, "click", sub_OnBtnUndo

    panel.style.textAlign = "left"
    panel.style.width = "640px"
    panel.style.margin = "0 auto"
    panel.style.padding = "5px"
    panel.style.fontFamily = "Arial, helvetica, sans-serif"
    panel.style.fontSize = ".85em"
    panel.style.border = "1px solid #666"
    panel.style.backgroundColor = "#333"
    panel.style.verticalAlign = "middle"

    InitToolList
End Sub
    
Sub InitToolList
    tool.style.marginRight = "15px"
    tool.style.padding = "5px"
    tool.style.verticalAlign = "top"
    DomCreate "option", tool, "Freehand"
    DomCreate "option", tool, "Line"
    DomCreate "option", tool, "Rectangle"
    DomCreate "option", tool, "Filled Rectangle"
    DomCreate "option", tool, "Circle"
    DomCreate "option", tool, "Filled Circle"
End Sub
    
Function SelectedColor    
    Dim r, g, b, c
    c = cp.value

    $If Javascript
        r = parseInt(c.substr(1,2), 16)
        g = parseInt(c.substr(3,2), 16)
        b = parseInt(c.substr(5,2), 16)
    $End If
        
    SelectedColor = _RGB(r, g, b)
End Function
```

**Comment by aurelVZ** [comment]
Very well dBox...i simply forget that github have discussion board

Aurel - basic4All

**Comment by boxgaming** [comment]
This example is a bit obsolete at this point.  The Dom methods have now been moved into their own module.  An up-to-date version can be found on the [Samples](https://github.com/boxgaming/qbjs/wiki/Samples) page.
